Add native support for networks represented by the SONATA format

nicolossus opened the pull request:
This PR adds support for building and simulating networks of point neurons represented by the SONATA format.

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+Overview of the SONATA format 
+-----------------------------
+
+The SONATA (Scalable Open Network Architecture TemplAte) format provides a framework for storage and exchange of 
```
```suggestion
The SONATA (Scalable Open Network Architecture Template) format provides a framework for storage and exchange of 
```

nicolossus replied:
Since SONATA is derived from **S**calable **O**pen **N**etwork **A**rchitecture **T**empl**A**te I think it makes sense to keep it as is. This is also in line with the spelling in the SONATA paper and documentation.

jessica-mitchell replied:
Ah, sorry you're right :)

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+The cell and synapse properties of the nodes and edges, respectively, can be described either individually or for 
+whole subsets. Nodes that share a global set of properties constitute a node type. Similarly, a subset of edges 
+that share a global set of properties constitute an edge type. Whether a property is stored individually or on a 
+per-type basis is up to the modeller. The number of node or edge types in a network model is typically small compared 
```
```suggestion
per-type basis is up to the modeler. The number of node or edge types in a network model is typically small compared 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+particular node or edge type can be easily accessed by its node or edge type id.
+
+Each node and edge in the network is explicitly tabulated in the binary format HDF5. Populations are hierarchically
+organized by utilizing HDF5 groups and datasets. The SONATA format requires certain HDF5 datasets, for instance 
```
```suggestion
organized by utilizing HDF5 groups and datasets. The SONATA format requires certain HDF5 datasets, for instance, 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+Each node and edge in the network is explicitly tabulated in the binary format HDF5. Populations are hierarchically
+organized by utilizing HDF5 groups and datasets. The SONATA format requires certain HDF5 datasets, for instance 
+a dataset containing all the node type ids. Properties that are 
+stored on an individual-basis, for instance synaptic weights, are also stored in HDF5 datasets. 
```
```suggestion
stored on an individual basis, for instance, synaptic weights, are also stored in HDF5 datasets. 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+
+.. _sec:sonata_nodes:
+
+The NEST support of SONATA nodes 
```
```suggestion
NEST support of SONATA nodes 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+.. _sec:sonata_nodes:
+
+The NEST support of SONATA nodes 
+--------------------------------
```
```suggestion
----------------------------
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+The NEST support of SONATA nodes 
+--------------------------------
+
+In the SONATA format, node populations are serialized in nodes HDF5 files and have a single associated node types 
```
```suggestion
In the SONATA format, node populations are serialized in node HDF5 files and have a single associated node type 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+--------------------------------
+
+In the SONATA format, node populations are serialized in nodes HDF5 files and have a single associated node types 
+CSV file that assigns properties to all nodes with a given node type id. A node types CSV file may be shared by 
```
```suggestion
CSV file that assigns properties to all nodes with a given node type id. A node type CSV file may be shared by 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+CSV file that assigns properties to all nodes with a given node type id. A node types CSV file may be shared by 
+multiple node population HDF5 files.
+
+NEST assumes the following structure of the nodes HDF5 files: 
```
```suggestion
NEST assumes the following structure of the node HDF5 files: 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+* ``virtual`` 
+
+Both ``point_neuron`` and ``point_process`` mean that the node is a neuron model (explicitly simulated) whereas ``virtual`` 
+means that the node only provide inputs to the simulated system. ``virtual`` nodes are modelled as ``spike_generator``\s 
```
```suggestion
means that the node only provide inputs to the simulated system. ``virtual`` nodes are modeled as ``spike_generator``\s 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+
+Both ``point_neuron`` and ``point_process`` mean that the node is a neuron model (explicitly simulated) whereas ``virtual`` 
+means that the node only provide inputs to the simulated system. ``virtual`` nodes are modelled as ``spike_generator``\s 
+(see :doc:`../models/spike_generator`\). NEST requires that only one ``model_type`` is present per node types CSV file. 
```
```suggestion
(see :doc:`../models/spike_generator`\). NEST requires that only one ``model_type`` is present per node type CSV file. 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+means that the node only provide inputs to the simulated system. ``virtual`` nodes are modelled as ``spike_generator``\s 
+(see :doc:`../models/spike_generator`\). NEST requires that only one ``model_type`` is present per node types CSV file. 
+
+The required headers for node types CSV files that describe neuron models are: 
```
```suggestion
The required headers for node type CSV files that describe neuron models are: 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+
+For a given ``node_type_id``, the ``model_template`` entry is the name of the NEST neuron model with prefix ``nest:``. NEST 
+does not require the ``model_template`` entries to be the same, but the creation of the nodes described in a single node 
+types CSV file is faster if the neuron models are the same. 
```
```suggestion
type CSV file is faster if the neuron models are the same. 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+NEST does not support node properties stored on an individual-basis in HDF5 datasets. This restriction can be 
+circumvented by assigning a single node its own node type id. 
+
+Below is an example of a node types CSV file with the required headers for neuron nodes: 
```
```suggestion
Below is an example of a node type CSV file with the required headers for neuron nodes: 
```

jessica-mitchell replied:
All of  these plural 'types' read awkwardly to me.  It seems like it should be singular descriptor.

nicolossus replied:
I tend to agree. I based the spelling on the SONATA documentation which use a plural descriptor, but see now that the SONATA paper use a singular.

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+| 2            | point_process | nest:iaf_psc_alpha | params_2.json   |
++--------------+---------------+--------------------+-----------------+
+
+The only required CSV header for ``virtual`` nodes is ``model_type``. The ``spike_generator``\s spike-times arrays are expected
```
```suggestion
The only required CSV header for ``virtual`` nodes is ``model_type``. The ``spike_generator``\s spike-time arrays are expected
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+The NEST support of SONATA edges  
+--------------------------------
+
+Analogous to nodes, edge populations are serialized in edges HDF5 files and have a single associated edge types 
```
```suggestion
Analogous to nodes, edge populations are serialized in edge HDF5 files and have a single associated edge types 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+Analogous to nodes, edge populations are serialized in edges HDF5 files and have a single associated edge types 
+CSV file that assigns properties to all edges with a given edge type id.
+
+NEST assumes the following structure of the edges HDF5 files: 
```
```suggestion
NEST assumes the following structure of the edge HDF5 files: 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+
+
+Together the ``source_node_id`` and ``target_node_id`` datasets explicitly tabulate all individual connections. 
+The ``edge_type_id`` dataset attributes each edge its edge type id which is used to assign synaptic properties from the 
```
```suggestion
The ``edge_type_id`` dataset attributes each edge its edge type id, which is used to assign synaptic properties from the 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+edge types CSV file. 
+
+In the SONATA format, edges within a population can be organized into one or more edge groups. Synaptic properties that 
+are specified on an individual-basis are stored in these edge groups. The groups are identified by an ``edge_id`` key. 
```
```suggestion
are specified on an individual basis are stored in these edge groups. The groups are identified by an ``edge_id`` key. 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+
+In the SONATA format, edges within a population can be organized into one or more edge groups. Synaptic properties that 
+are specified on an individual-basis are stored in these edge groups. The groups are identified by an ``edge_id`` key. 
+NEST assumes the ``edge_id``\s are contiguous numeric keys starting from zero, i.e. 0, 1, 2, ... 
```
```suggestion
NEST assumes the ``edge_id``\s are contiguous numeric keys starting from zero, that is, 0, 1, 2, ... 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+are specified on an individual-basis are stored in these edge groups. The groups are identified by an ``edge_id`` key. 
+NEST assumes the ``edge_id``\s are contiguous numeric keys starting from zero, i.e. 0, 1, 2, ... 
+
+**Note:** NEST currently only supports one edge group per edge population. Furthermore, NEST only reads the ``delay`` 
```
Use the Restructured Text format for notes

```
.. note::

     NEST currently  . . .

```

nicolossus replied:
Resolved by [3941e90](https://github.com/nest/nest-simulator/pull/2595/commits/3941e90f64518f2c102252ea7742b1bf5bd0bc6d).

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+    │  ├─ <population_name>             Group - required - usually only one but can be more population groups per file
+    │  │  ├─ node_type_id               Dataset {N_total_nodes} - required
+
+**Note:** NEST assumes that the implicit row numbers in the ``node_type_id`` dataset correspond to the ``node_id``\s. 
```
Use the Restructuctured format for notes:

```
.. note::

   NEST assumes . . .
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+
+**Note:** NEST currently only supports one edge group per edge population. Furthermore, NEST only reads the ``delay`` 
+and ``syn_weight`` datasets, given that they are provided. This means that only axonal delays and synaptic weights can 
+be stored on an individual-basis in the HDF5 format. Other synaptic properties must be given in the edge types 
```
```suggestion
be stored on an individual basis in the HDF5 format. Other synaptic properties must be given in the edge type 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+be stored on an individual-basis in the HDF5 format. Other synaptic properties must be given in the edge types 
+CSV file(s). 
+
+Below is an example of a edge types CSV file: 
```
```suggestion
Below is an example of a edge type CSV file: 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+| 2            | static_synapse | 2.5   | params_2.json   |
++--------------+----------------+-------+-----------------+
+
+**Note:** Only the synaptic properties ``delay`` and ``syn_weight`` can be provided as headers in the edge types CSV file. 
```
Use restructured text format for notes (see above for example)

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+      }
+    }
+
+**Note:** NEST supports the use of two config files, i.e. one network and one simulation config. NEST does not 
```
Use restructured text format for notes

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+    }
+
+
+NEST does not support node properties stored on an individual-basis in HDF5 datasets. This restriction can be 
```
```suggestion
NEST does not support node properties stored on an individual basis in HDF5 datasets. This restriction can be 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+The SONATA configuration files
+------------------------------
+
+Model metadata, such as the relative location of the network files, and simulation parameters are stored in the 
```
```suggestion
Model metadata, such as the relative location of the network files and simulation parameters, are stored in the 
```

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,259 @@
+
+Model metadata, such as the relative location of the network files, and simulation parameters are stored in the 
+SONATA configuration ("config") file(s) in the JSON format. Below is an example SONATA config with the components NEST 
+expects included:
```
```suggestion
expects to be included:
```

jessica-mitchell reviewed (changes requested):
Thanks @nicolossus I looked over the documentation and have some minor changes.

clinssen commented on nestkernel/sonata_connector.cpp:
```diff
@@ -0,0 +1,667 @@
+// Includes from sli:
+#include "dictutils.h"
+
+#include <chrono>   // for debugging
```
Can we use a debug flag from cmake here?

nicolossus replied:
I just removed the includes as they were not used anywhere.

clinssen commented on nestkernel/sonata_connector.cpp:
```diff
@@ -0,0 +1,667 @@
+#include <iostream> // for debugging
+#include <stdio.h>  // for debugging
+
+#include "H5Cpp.h" // HDF5 C++ API
```
```suggestion
#include "H5Cpp.h"
```

clinssen commented on nestkernel/sonata_connector.cpp:
```diff
@@ -0,0 +1,667 @@
+      const auto pop_grp = open_group_( edges_grp, pop_name );
+
+      // Find the number of edge id groups and edge id group names
+      // NOTE: current find_edge_id_groups_() is only meant as a temporary helper function
```
can we move this comment to the docstring for ``find_edge_id_groups_()``?

clinssen commented on nestkernel/sonata_connector.cpp:
```diff
@@ -0,0 +1,667 @@
+void
+SonataConnector::try_open_edge_group_id_dsets_( const H5::Group* edge_id_grp )
+{
+  // TODO: Currently only works if the edge file has a single edge id group
```
Could we assert() this here?

nicolossus replied:
We currently verify that only one edge group is present before this function is called:

https://github.com/nicolossus/nest-simulator/blob/689502d40e6cfc9b8fd4bcdace83db1d5b3bbbef/nestkernel/sonata_connector.cpp#L128-L132

In its current implementation, the function only receives a pointer to a single edge group. I'm therefore not sure if we should also do an assert here. We would then need to pass the number of found edge groups to the function, which already should have raised an error if there are more than one.

The comment is more directed towards that we need to rework this function when we add support for more than
one edge group.

clinssen commented on nestkernel/sonata_connector.cpp:
```diff
@@ -0,0 +1,667 @@
+hsize_t
+SonataConnector::find_edge_id_groups_( H5::Group* pop_grp, std::vector< std::string >& edge_id_grp_names )
+{
+  // Find the number of edge id groups, i.e. ones with label "0", "1", ..., by finding
```
Can this comment block be moved to the method's docstring?

clinssen commented on nestkernel/sonata_connector.cpp:
```diff
@@ -0,0 +1,667 @@
+  // create it here once to avoid re-creating the object over and over again.
+  for ( thread tid = 0; tid < kernel().vp_manager.get_num_threads(); ++tid )
+  {
+    type_id_2_syn_spec_[ type_id ].push_back( synapse_params ); // DO WE NEED TO DEFINE THIS PER THREAD???
```
This comment looks like one that needs to be resolved before merge.

nicolossus replied:
Thanks for pointing this out, the comment is now resolved. Please take a look at the new implementation.

clinssen commented on pynest/nest/lib/hl_api_sonata.py:
```diff
@@ -0,0 +1,615 @@
+            # Recursive substitutions of path variables with entries from manifest
+            if isinstance(config_obj, dict):
+                return {k: recursive_substitutions(v) for k, v in config_obj.items()}
+            elif isinstance(config_obj, list):
```
You can use the return-early pattern here, "else" is not necessary because all of the cases return from the function.

clinssen commented on pynest/nest/lib/hl_api_sonata.py:
```diff
@@ -0,0 +1,615 @@
+    def Create(self):
+        """Create the SONATA network nodes.
+
+        The network nodes are created on the Python level. In the SONATA format,
```
What does "The network nodes are created on the Python level" mean here? Do we create nodes one-by-one, or instantiate a whole population at once on the C++ level?

nicolossus replied:
I removed the "on the Python level" because the nodes are of course ultimately created by the kernel. I think what was meant here is that we do all SONATA handling for node creation on the Python level, which is opposed to the handling of edges where most of the SONATA handling is done by the kernel.

clinssen commented on testsuite/pytests/test_sonata/300_pointneurons/network/internal_node_types.csv:
```diff
@@ -0,0 +1,6 @@
+node_type_id ei model_template model_type dynamics_params model_name
+104 i nest:iaf_psc_alpha point_process 473862421_point.json PV2
+100 e nest:iaf_psc_alpha point_process 472363762_point.json Scnn1a
```
Are all these files duplicated between ``examples`` and ``pytests``? Can we use a symbolic link or some other method of de-duplication (cc @jessica-mitchell)?

nicolossus replied:
While I agree it would be best to avoid duplication, I don't think there is a simple solution for this. The SONATA files should reside in the examples dir in order to be accessible to users. I tried to create a symlink in the testsuite dir pointing to the files in the examples dir. This doesn't work as the files are not found when, for instance, running `installcheck` from a build dir. The files comprise about 2.6 MB, so I don't think duplication matters too much when it comes to file size.

clinssen reviewed (commented):
Thanks for this PR, the code looks very good! I just have a few minor comments.

jessica-mitchell reviewed (approved):
thanks @nicolossus looks good from my perspective

heplesser commented on nestkernel/connection_manager.cpp:
```diff
@@ -32,6 +32,7 @@
 #include <iomanip>
 #include <limits>
 #include <set>
+#include <string>
```
Why is this needed?

nicolossus replied:
Just a remnant from debugging, removed.

heplesser commented on nestkernel/connection_manager.h:
```diff
@@ -158,6 +158,8 @@ class ConnectionManager : public ManagerInterface
     size_t n,
     std::string syn_model );
 
+  void connect_sonata( const DictionaryDatum& graph_specs, const long chunk_size );
```
Please add a doxygen comment explaining what dictionary is expected and what chunk_size is.

heplesser commented on CMakeLists.txt:
```diff
@@ -59,6 +59,7 @@ set( with-libneurosim OFF CACHE STRING "Build with libneurosim [default=OFF]. Op
 set( with-music OFF CACHE STRING "Build with MUSIC [default=OFF]. Optionally give the directory where MUSIC is installed." )
 set( with-sionlib OFF CACHE STRING "Build with SIONlib [default=OFF]. Optionally give the directory where sionlib is installed." )
 set( with-boost ON CACHE STRING "Build with Boost [default=ON]. To set a specific Boost installation, give the install path." )
+set( with-sonata ON CACHE STRING "Find a HDF5 library. To set a specific HDF5 installation, set install path. [default=ON]" )
```
This option should be `with-hdf5` because it controls if and optionally which HDF5 library is included. We may later use HDF5 in other contexts than Sonata and then it will become confusing to have the option named for Sonata.

nicolossus replied:
I have now replaced option names and messages to use HDF5 instead of SONATA.

heplesser reviewed (commented):
@nicolossus Thank you for this PR! I have a few small comments, see in the review. In addition, I am concerned about the duplication of the large number of data files (json, csv, h5) involved in example and test, particularly as the h5 files are 1 MB in size. Can we use the files on the example for the tests?

heplesser commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,268 @@
+* ``point_process``
+* ``virtual`` 
+
+Both ``point_neuron`` and ``point_process`` mean that the node is a neuron model (explicitly simulated) whereas ``virtual`` 
```
```suggestion
Both ``point_neuron`` and ``point_process`` mean that the node is a neuron model (explicitly simulated; the two terms can be used interchangeably) whereas ``virtual`` 
```

med-ayssar commented:
@heplesser and @nicolossus what is the current status?

heplesser commented:
@steffengraber If your tests with Conda packaging of this branch were successful, could you approve the PR?

@nicolossus Can you fix the conflict?

@jougs Could you instead of @terhorstd look at the packaging/release aspect, see my question above?

steffengraber reviewed (commented):
@nicolossus With the most recent changes from April 14 and 15, I can no longer build a Conda package. Before that, it worked. 

As an example the linux python3.10 package:

```
[ 1%] Building CXX object libnestutil/CMakeFiles/nestutil.dir/stopwatch.cpp.o
[2%] Linking CXX static library libnestutil.a
Error running link command: No such file or directory
make[2]: *** [libnestutil/CMakeFiles/nestutil.dir/build.make:146: libnestutil/libnestutil.a] Error 2
make[1]: *** [CMakeFiles/Makefile2:488: libnestutil/CMakeFiles/nestutil.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

```

The complete error log can be found here: https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/691087/logs/37

The recipe source can be found here: https://github.com/steffengraber/nest-simulator-feedstock/tree/test-nestsonata

I have already tried a few things. Do you have any idea what the problem could be?

nicolossus commented:
@steffengraber This might be related to https://github.com/nest/nest-simulator/pull/2659. I have merged with the most recent master. Could you try to build the conda package again?

jougs commented on cmake/ConfigureSummary.cmake:
I'm not sure I like the mix of reformatting and code changes. It is really hard to see what has actually changed and this makes code inconsistent with respect to many other CMake files. Can you please revert the formatting changes?

jougs commented on pynest/nest/lib/hl_api_sonata.py:
```diff
@@ -0,0 +1,651 @@
+    def __init__(self, config, sim_config=None):
+
+        if not have_hdf5:
+            msg = ("SonataNetwork unavailable because NEST was compiled "
+                   "without HDF5 support")
```
I don't think this line should be broken. Same applies below.

jougs reviewed (changes requested):
Here's some comments from the release/CI perspective. I did not check the code or documentation. Please tell me if I should do that.

steffengraber commented:
> @steffengraber This might be related to #2659. I have merged with the most recent master. Could you try to build the conda package again?

@nicolossus I have restarted build. But same problem: https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/692105/logs/41

heplesser commented:
> > @steffengraber This might be related to #2659. I have merged with the most recent master. Could you try to build the conda package again?
> 
> @nicolossus I have restarted build. But same problem: https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/692105/logs/41

@steffengraber Can you check if you can build a conda package with current master? This seems likely unrelated to this part, but related to the recently merged #2655.

nicolossus commented on build_support/parse_build_log.py:
```diff
@@ -321,6 +321,9 @@ def makebuild_summary(log_filename, msg_make_section_start,
     # If a certain build_type (e.g. 'MINIMAL' or 'FULL') only builds
     # with some warnings, this would be a good point to re-set the
     # expected_warnings variable conditionally for that build_type.
+    # NOTE: The sonata_connector.cpp warnings are due to a bug in
+    # HDF 1.10.4 and has been fixed in later versions. However, only
+    # HDF 1.10.4 is available for Ubuntu 20.04, which the CI uses.
```
The default `ubuntu-latest` for [GitHub Actions is now Ubuntu 22.04](https://github.blog/changelog/2022-11-09-github-actions-ubuntu-latest-workflows-will-use-ubuntu-22-04/), so the time for change seems good. However, this should probably be in a separate PR.

jougs replied:
Definitely separate PR, yes. Thanks for checking. Maybe we could look into this next week?

nicolossus replied:
Sounds good!

heplesser commented on CMakeLists.txt:
```diff
@@ -59,6 +59,7 @@ set( with-libneurosim OFF CACHE STRING "Build with libneurosim [default=OFF]. Op
 set( with-music OFF CACHE STRING "Build with MUSIC [default=OFF]. Optionally give the directory where MUSIC is installed." )
 set( with-sionlib OFF CACHE STRING "Build with SIONlib [default=OFF]. Optionally give the directory where sionlib is installed." )
 set( with-boost ON CACHE STRING "Build with Boost [default=ON]. To set a specific Boost installation, give the install path." )
+set( with-hdf5 ON CACHE STRING "Find a HDF5 library. To set a specific HDF5 installation, set install path. [default=ON]" )
```
Good point.

nicolossus commented:
@jougs I think I have addressed your comments. I also added some documentation on how to configure NEST for SONATA. @heplesser Can you also see if you are happy with the addition to the documentation?

heplesser commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -4,8 +4,26 @@ NEST SONATA guide
+
+.. code-block:: sh
+
+    cmake -Dwith-hdf5=[ON </path/to/hdf5>]
```
- I find this line misleading since it looks like a complete cmake invocation, but isn't. I would rather write that you need to add the `-Dwith-hdf5=ON` option to the cmake invocation. @jessica-mitchell What do you think?
- Does `/path/to/hdf5` really work?

jessica-mitchell replied:
I agree, I think incomplete commands can be confusing.

heplesser commented on doc/htmldoc/installation/cmake_options.rst:
```diff
@@ -87,6 +87,9 @@ External libraries
 +-------------------------------------------------------+------------------------------------------------------------------------------------------------+
++-------------------------------------------------------+------------------------------------------------------------------------------------------------+
+| ``-Dwith-hdf5=[OFF|ON|</path/to/hdf5>]``              | Build with HDF5 library [default=OFF]. To set a specific library, give the install path.       |
++-------------------------------------------------------+------------------------------------------------------------------------------------------------+
```
Here (and also for GSL above and probably other options), we only explain technically how to build NEST with a given library. But we say nothing about the purpose or actual effect. Should we mention here that HDF5 is needed for Sonata, and similarly for GSL that it is needed for a range of neuron models? @jessica-mitchell @jougs Comments?

nicolossus replied:
I suggest we improve this in a separate PR. I have created issue https://github.com/nest/nest-simulator/issues/2765 to remind us of the needed improvements.

jessica-mitchell replied:
Yes I think this makes sense to add more context. We could, in my opinion, add some relevant text here for SONATA, and in a separate PR look at adding more context to all the options.

heplesser reviewed (commented):
Please see my inline comment.s

jessica-mitchell commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,302 @@
+
+    -Dwith-hdf5=ON
+
+For further details, see :doc:`../installation/cmake_options`.
```
This should be :ref:`cmake_options`

nicolossus commented:
I have merged with the current master and done the necessary updates. 

@heplesser and @jessica-mitchell The documentation on configuring NEST for SONATA in `nest_sonata_guide.rst` is updated and I also added some context for the HDF5 option in `cmake_options.rst`. Do you think the documentation looks fine for now?

@jougs I have adapted the build matrix to also include HDF5 for the full NEST build. Can you check that the changes are okay?

jessica-mitchell reviewed (changes requested):
@nicolossus Looks good, just noticed that the link to the example is broken because the path is incorrect (see comments), also you need to a README to the pynest/example/sonata_example/ page for sphinx-gallery to generate output from subdirectories.

nicolossus commented:
@jessica-mitchell Thanks, fixed your comments.

jessica-mitchell reviewed (approved):
Great, looks good!

heplesser commented on doc/htmldoc/installation/cmake_options.rst:
```diff
@@ -87,6 +87,11 @@ External libraries
 | ``-Dwith-gsl=[OFF|ON|</path/to/gsl>]``                | Build with the GSL library [default=ON]. To set a specific library, give the install path.     |
 +-------------------------------------------------------+------------------------------------------------------------------------------------------------+
++-------------------------------------------------------+------------------------------------------------------------------------------------------------+
+| ``-Dwith-hdf5=[OFF|ON|</path/to/hdf5>]``              | Build with `HDF5 <https://hdfgroup.org/>`_ library [default=OFF]. To set a specific library,   |
+|                                                       | give the install path. HDF5 is needed to configure NEST for                                    |
+|                                                       | `SONATA <https://github.com/AllenInstitute/sonata>`_ support.                                  |
```
```suggestion
| ``-Dwith-hdf5=[OFF|ON|</path/to/hdf5>]``              | Build with `HDF5 <https://hdfgroup.org/>`_ library [default=OFF]. To set a specific library,   |
|                                                       | give the install path. HDF5 is required for                                    |
|                                                       | `SONATA <https://github.com/AllenInstitute/sonata>`_ support.                                  |
```

I think it would be better to link to `nest_sonata_guide.rst` here instead of to the generic sonata homepage. That will lead the user to more relevant documentation.

heplesser commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,302 @@
+NEST SONATA guide 
+=====================
+
+NEST supports building and simulating networks of point neurons described by the SONATA format [1]_. 
```
Could we put a direct link to the Sonata homepage in here?

heplesser commented on pynest/examples/sonata_example/300_pointneurons/circuit_config.json:
```diff
@@ -0,0 +1,35 @@
+    ]
+  },
+  "target_simulator": "NEST"
+}
```
Newline missing at end.

heplesser commented on pynest/nest/lib/hl_api_sonata.py:
```diff
@@ -0,0 +1,693 @@
+        self._edges_maps = []
+        self._chunk_size_default = 2**20
+
+        self._is_nodes_created = False
```
```suggestion
        self._are_nodes_created = False
```

heplesser commented on pynest/nest/lib/hl_api_sonata.py:
```diff
@@ -0,0 +1,693 @@
+
+                for row in edges_df.itertuples(name=None):
+                    # Set of settable parameters
+                    settable_params = set([*GetDefaults(row[idx_map["synapse_model"]])])
```
The code following here looks very similar to the code in the `if is_one_model` case. The entire method here is also very long. Could you extract the code in the `if` and (most of) the `else` clauses into separate method(s) and maybe even avoid code duplication?

heplesser commented on nestkernel/sonata_connector.cpp:
```diff
@@ -0,0 +1,666 @@
+    }
+  }
+
+  // TODO: If present, ensure correct size of syn_weight and delay dsets. This might not be straightforward if there
+  // are multiple edge id groups
```
This is not an issue as long as we only support a single edge group, or? Then I would remove this comment, since it gives the impression that the current implementation is not safe. Are the sizes currently checked elsewhere?

heplesser reviewed (changes requested):
@nicolossus This looks very fine, I just have a few details, see inline.

heplesser commented on doc/htmldoc/installation/cmake_options.rst:
```diff
@@ -87,6 +87,11 @@ External libraries
++-------------------------------------------------------+------------------------------------------------------------------------------------------------+
+| ``-Dwith-hdf5=[OFF|ON|</path/to/hdf5>]``              | Build with `HDF5 <https://hdfgroup.org/>`_ library [default=OFF]. To set a specific library,   |
+|                                                       | give the install path. HDF5 is required for                                                    |
+|                                                       | `SONATA <https://github.com/AllenInstitute/sonata>`_ support.                                  |
```
Just one last little thing: I think this link should go to our `nest_sonata_guide.rst` to provide more context for the user. The general Sonata home page will not help to make the connection to this compiler option.

heplesser reviewed (approved):
@nicolossus Thanks a lot, looks good to me and the failing on macOS seems to be a runner problem, so I will approve now, assuming that the test will pass on re-run.

jougs commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,303 @@
+.. _nest_sonata:
+
+NEST SONATA guide 
+=====================
```
```suggestion
=================
```

jougs commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,303 @@
+.. _sec:sonata_configure:
+
+Configure NEST for SONATA
+--------------------------
```
```suggestion
-------------------------
```

jougs commented on doc/htmldoc/nest_sonata/nest_sonata_guide.rst:
```diff
@@ -0,0 +1,303 @@
+.. _sec:sonata_examples:
+
+NEST SONATA Example
+--------------------
```
```suggestion
-------------------
```
